src/runtime_src/xdp/profile/database/static_info/device_info.cpp

```diff
@@ -39,260 +38,316 @@ namespace xdp {
+      xclbinAvailable = true;
+    
+    if (!xclbinAvailable) {
+      for(auto xclbin : loadedConfigInfos.back()->currentXclbins) {
+        if(xclbin->type == xclbinQueryType) {
+          xclbinAvailable = true;
+          break;
+        }
+      }
+    }
+
+    if(xclbinAvailable) {
+      xrt_core::message::send(xrt_core::message::severity_level::info, "XRT", "Missing xclbin is available in config.");
+      ConfigInfo* lastCfg = loadedConfigInfos.back().get();
+      for(auto &xclbin : lastCfg->currentXclbins)
+      {
+        if(xclbin->type == xclbinQueryType || xclbin->type == XCLBIN_AIE_PL) {
+          // Create a copy of required missing xclbinInfo.
+          requiredXclbinInfo = new XclbinInfo(xclbinQueryType);
+          if(xclbinQueryType == XCLBIN_AIE_ONLY)
+          {
+            // Copying missing AIE xclbin
+            requiredXclbinInfo->aie = xclbin->aie;
+            requiredXclbinInfo->pl.valid = false ;
+          }
+          else
+          {
+            // Copying missing PL xclbin
+            requiredXclbinInfo->pl = xclbin->pl;
+            requiredXclbinInfo->aie.valid = false ;
+
+            // TODO : Update & delete deviceIntf from previous config
+            // requiredXclbinInfo->deviceIntf = new DeviceIntf();
+            // delete xclbin->deviceIntf;
+            // xclbin->deviceIntf = nullptr;
+
+          }
+          requiredXclbinInfo->uuid = xclbin->uuid ;
+          requiredXclbinInfo->name = xclbin->name ;
+          break;  // Need only one such missing xclbinInfo from last config.
+        }
       }
     }
-    loadedXclbins.push_back(xclbin) ;
+    return requiredXclbinInfo;
+  }
+
+  void DeviceInfo::createConfig(XclbinInfo* xclbin)
+  {
+    // Check if loaded history has same UUID.
+    if (!loadedConfigInfos.empty() && loadedConfigInfos.back()->containsXclbin(xclbin->uuid)) {
+      return;
+    }
+
+    // Create a new config
+    std::unique_ptr<ConfigInfo> config = std::make_unique<ConfigInfo>();
+    config->addXclbin(xclbin);
+
+    auto currentXclbinType = xclbin->type;
+
+    // Check if this itself is a complete xclbin (AIE+PL).
+    if (currentXclbinType == XCLBIN_AIE_PL)
+    {
+      loadedConfigInfos.push_back(std::move(config));
+      return;
+    }
+
+    // If it is not a complete xclbin.
+    //  Check what is missing & request that missing XclbinInfo.?
+    //    a. AIEInfo or PLInfo 
+    XclbinInfo *missingXclbin = nullptr;
+    if (currentXclbinType == XCLBIN_AIE_ONLY)
+    {
+      xclbin->pl.valid = false ;
+      missingXclbin = createXclbinFromLastConfig(XCLBIN_PL_ONLY);
+    }
+    else
+    {
+      xclbin->aie.valid = false ;
+      missingXclbin = createXclbinFromLastConfig(XCLBIN_AIE_ONLY);
+    }
+
+    // If missing part of XclbinInfo is available. 
+    if(missingXclbin)
+    {
+      config->currentXclbins.back()->aie.numTracePLIO = loadedConfigInfos.size() == 0 ? 0 : loadedConfigInfos.back()->currentXclbins.back()->aie.numTracePLIO;
```

Comment:
Is this safe if we load pl xclbin first? Why do we have to set this?

Reply:
Somehow github doesn't show responses I had posted to review. Re-posting
"As part of initializing profile monitors,and hence ts2mm initialization from PL Xclbin, number of trace PLIO gets recorded in AIEInfo. Hence to account that information, this check is added.
Ideally we can move number of trace PLIO information to config, I can take care of that in follow-up commit."